# Session 17 Pandas DataFrame

In [1]:
import pandas as pd
import numpy as np

## 📘 DataFrame (in Pandas)

A **DataFrame** is a **two-dimensional (2D)**, **labeled data structure** provided by the **`pandas`** library in Python. It is one of the most powerful and essential tools in data science for working with **tabular data**, similar to how data is stored in **Excel spreadsheets** or **SQL tables**.

### ✅ Key Characteristics:

* Consists of **rows and columns**
* Each **column** can contain different **data types** (integers, floats, strings, etc.)
* Both **rows (index)** and **columns (labels)** are **labeled**, making it easy to select and manipulate data
* Ideal for data cleaning, analysis, and manipulation tasks

---

## 📐 2D Structure of DataFrame

A DataFrame is a **2D object**, meaning:

* **Rows** run horizontally (axis=0)
* **Columns** run vertically (axis=1)

### Example Table Representation:

```
DataFrame
  ┌────────────┬─────┬────────────┐
  │ Name       │ Age │ City       │  → Columns
  ├────────────┼─────┼────────────┤
0 │ Alice      │ 25  │ Delhi      │
1 │ Bob        │ 30  │ Mumbai     │
2 │ Charlie    │ 22  │ Kolkata    │
  └────────────┴─────┴────────────┘
     ↑
   Index (Rows)
```

---

## 📊 Role of Series in a DataFrame

The **`Series`** is the **building block** of a DataFrame.

* A **Series** is a **1D labeled array**.
* In a DataFrame:

  * Each **column** is a **Series** with the column name as the label and row indices as the index.
  * Each **row** is also treated as a **Series**, where the column labels serve as the index.

### Conceptual View:

* `DataFrame = Collection of Series objects (columns)`
* `Row = Series where index = column names`
* `Column = Series where index = row numbers`

---

## 💻 Example:

```python
import pandas as pd

# Create a DataFrame from a dictionary
data = {
    'Name': ['Alice', 'Bob', 'Charlie'],
    'Age': [25, 30, 22],
    'City': ['Delhi', 'Mumbai', 'Kolkata']
}

df = pd.DataFrame(data)

# Display the DataFrame
print(df)

# Access a column (returns a Series)
print(type(df['Age']))      # Output: <class 'pandas.core.series.Series'>

# Access a row (also returns a Series)
print(type(df.loc[0]))      # Output: <class 'pandas.core.series.Series'>
```

---

### 📌 Summary Table

| Feature          | Description                                   |
| ---------------- | --------------------------------------------- |
| **2D Structure** | Rows (axis=0) and Columns (axis=1)            |
| **Column**       | A `pandas.Series` object                      |
| **Row**          | Also behaves as a `pandas.Series`             |
| **Use Case**     | Essential in data analysis, cleaning, ML prep |

## Methods to read/create a Dataframe
Provided below are the methods using which Dataframes can either be read or created.

In [2]:
# creating df using a list
stud_list = [
    [100, 80, 10],
    [90, 70, 7],
    [120, 100, 14],
    [80, 5, 2],
    [100, 80, 10]
]

stud = pd.DataFrame(stud_list, columns=['iq', 'marks', 'pkg'])

In [3]:
# using dictionaries
stud_dict = {
    'name': ['nool', 'wool', 'rupaid', 'shabh', 'edit', 'amnit'],
    'iq': [100, 90, 120, 80, 0, 0],
    'marks' : [80, 70, 100, 50, 0, 0],
    'pkg': [10, 7, 14, 2, 0, 0]
}

students = pd.DataFrame(stud_dict)
students

name   iq  marks  pkg
0    nool  100     80   10
1    wool   90     70    7
2  rupaid  120    100   14
3   shabh   80     50    2
4    edit    0      0    0
5   amnit    0      0    0

In [2]:
# using read_csv
movies = pd.read_csv('movies.csv')
movies

title_x     imdb_id  \
0                 Uri: The Surgical Strike   tt8291224   
1                            Battalion 609   tt9472208   
2     The Accidental Prime Minister (film)   tt6986710   
3                          Why Cheat India   tt8108208   
4                          Evening Shadows   tt6028796   
...                                    ...         ...   
1624                 Tera Mera Saath Rahen   tt0301250   
1625                  Yeh Zindagi Ka Safar   tt0298607   
1626                       Sabse Bada Sukh   tt0069204   
1627                                 Daaka  tt10833860   
1628                              Humsafar   tt2403201   

                                            poster_path  \
0     https://upload.wikimedia.org/wikipedia/en/thum...   
1                                                   NaN   
2     https://upload.wikimedia.org/wikipedia/en/thum...   
3     https://upload.wikimedia.org/wikipedia/en/thum...   
4                                                   NaN   
...                                                 ...   
1624  https://upload.wikimedia.org/wikipedia/en/2/2b...   
1625  https://upload.wikimedia.org/wikipedia/en/thum...   
1626                                                NaN   
1627  https://upload.wikimedia.org/wikipedia/en/thum...   
1628  https://upload.wikimedia.org/wikipedia/en/thum...   

                                              wiki_link  \
0     https://en.wikipedia.org/wiki/Uri:_The_Surgica...   
1           https://en.wikipedia.org/wiki/Battalion_609   
2     https://en.wikipedia.org/wiki/The_Accidental_P...   
3         https://en.wikipedia.org/wiki/Why_Cheat_India   
4         https://en.wikipedia.org/wiki/Evening_Shadows   
...                                                 ...   
1624  https://en.wikipedia.org/wiki/Tera_Mera_Saath_...   
1625  https://en.wikipedia.org/wiki/Yeh_Zindagi_Ka_S...   
1626      https://en.wikipedia.org/wiki/Sabse_Bada_Sukh   
1627                https://en.wikipedia.org/wiki/Daaka   
1628             https://en.wikipedia.org/wiki/Humsafar   

                            title_y                 original_title  is_adult  \
0          Uri: The Surgical Strike       Uri: The Surgical Strike         0   
1                     Battalion 609                  Battalion 609         0   
2     The Accidental Prime Minister  The Accidental Prime Minister         0   
3                   Why Cheat India                Why Cheat India         0   
4                   Evening Shadows                Evening Shadows         0   
...                             ...                            ...       ...   
1624          Tera Mera Saath Rahen          Tera Mera Saath Rahen         0   
1625           Yeh Zindagi Ka Safar           Yeh Zindagi Ka Safar         0   
1626                Sabse Bada Sukh                Sabse Bada Sukh         0   
1627                          Daaka                          Daaka         0   
1628                       Humsafar                       Humsafar         0   

      year_of_release runtime            genres  imdb_rating  imdb_votes  \
0                2019     138  Action|Drama|War          8.4       35112   
1                2019     131               War          4.1          73   
2                2019     112   Biography|Drama          6.1        5549   
3                2019     121       Crime|Drama          6.0        1891   
4                2018     102             Drama          7.3         280   
...               ...     ...               ...          ...         ...   
1624             2001     148             Drama          4.9         278   
1625             2001     146             Drama          3.0         133   
1626             2018      \N      Comedy|Drama          6.1          13   
1627             2019     136            Action          7.4          38   
1628             2011      35     Drama|Romance          9.0        2968   

                                              

In [3]:
ipl = pd.read_csv('ipl-matches.csv')
ipl

ID        City        Date   Season  MatchNumber  \
0    1312200   Ahmedabad  2022-05-29     2022        Final   
1    1312199   Ahmedabad  2022-05-27     2022  Qualifier 2   
2    1312198     Kolkata  2022-05-25     2022   Eliminator   
3    1312197     Kolkata  2022-05-24     2022  Qualifier 1   
4    1304116      Mumbai  2022-05-22     2022           70   
..       ...         ...         ...      ...          ...   
945   335986     Kolkata  2008-04-20  2007/08            4   
946   335985      Mumbai  2008-04-20  2007/08            5   
947   335984       Delhi  2008-04-19  2007/08            3   
948   335983  Chandigarh  2008-04-19  2007/08            2   
949   335982   Bangalore  2008-04-18  2007/08            1   

                           Team1                        Team2  \
0               Rajasthan Royals               Gujarat Titans   
1    Royal Challengers Bangalore             Rajasthan Royals   
2    Royal Challengers Bangalore         Lucknow Super Giants   
3               Rajasthan Royals               Gujarat Titans   
4            Sunrisers Hyderabad                 Punjab Kings   
..                           ...                          ...   
945        Kolkata Knight Riders              Deccan Chargers   
946               Mumbai Indians  Royal Challengers Bangalore   
947             Delhi Daredevils             Rajasthan Royals   
948              Kings XI Punjab          Chennai Super Kings   
949  Royal Challengers Bangalore        Kolkata Knight Riders   

                                          Venue                   TossWinner  \
0              Narendra Modi Stadium, Ahmedabad             Rajasthan Royals   
1              Narendra Modi Stadium, Ahmedabad             Rajasthan Royals   
2                         Eden Gardens, Kolkata         Lucknow Super Giants   
3                         Eden Gardens, Kolkata               Gujarat Titans   
4                      Wankhede Stadium, Mumbai          Sunrisers Hyderabad   
..                                          ...                          ...   
945                                Eden Gardens              Deccan Chargers   
946                            Wankhede Stadium               Mumbai Indians   
947                            Feroz Shah Kotla             Rajasthan Royals   
948  Punjab Cricket Association Stadium, Mohali          Chennai Super Kings   
949                       M Chinnaswamy Stadium  Royal Challengers Bangalore   

    TossDecision SuperOver                  WinningTeam    WonBy  Margin  \
0            bat         N               Gujarat Titans  Wickets     7.0   
1          field         N             Rajasthan Royals  Wickets     7.0   
2          field         N  Royal Challengers Bangalore     Runs    14.0   
3          field         N               Gujarat Titans  Wickets     7.0   
4            bat         N                 Punjab Kings  Wickets     5.0   
..           ...       ...                          ...      ...     ...   
945          bat         N        Kolkata Knight Riders  Wickets     5.0   
946          bat         N  Royal Challengers Bangalore  Wickets     5.0   
947          bat         N             Delhi Daredevils  Wickets     9.0   
948          bat         N          Chennai Super Kings     Runs    33.0   
949        field         N        Kolkata Knight Riders     Runs   140.0   

    method Player_of_Match                                       Team1Players  \
0      NaN       HH Pandya  ['YBK Jaiswal', 'JC Buttler', 'SV Samson', 'D ...   
1      NaN      JC Buttler  ['V Kohli', 'F du Plessis', 'RM Patidar', 'GJ ...   
2      NaN      RM Patidar  ['V Kohli', 'F du Plessis', 'RM Patidar', 'GJ ...   
3      NaN       DA Miller  ['YBK Jaiswal', 'JC Buttler', 'SV Samson', 'D ...   
4      NaN   Harpreet Brar  ['PK Garg', 'Abhishek Sharma', 'RA Tripathi', ...   
..     ...             ...                                                ...   
945    NaN       DJ Hussey  ['WP Saha', 'BB

## DataFrame Attributes and Methods
Up ahead, we will be discussing some of the **Attributes** and **Methods** that are often used upon **DataFrames**. Some of the crucial ones are discussed below:

### Shape
Provides the shape of the datafrane, which is typically the *size*, *dimensions*, or the *total number* of **rows**, **columns** and **tables** (in case of a **multidim DF**)

In [6]:
print(movies.shape)
print(ipl.shape)

(1629, 18)
(950, 20)


### Dtypes
Provides the **Data Type** of the DatFrame. Since a DF comprises of *multiple series*, and each series can have either **heterogeneous** or **homogeneous** types of data, each series is labelled with a Data Type. The provide us with the insights regarding the dimensions and the size of data.

A series with *heterogeneous* types of data is marked as an **Object**.

In [7]:
movies.dtypes

title_x              object
imdb_id              object
poster_path          object
wiki_link            object
title_y              object
original_title       object
is_adult              int64
year_of_release       int64
runtime              object
genres               object
imdb_rating         float64
imdb_votes            int64
story                object
summary              object
tagline              object
actors               object
wins_nominations     object
release_date         object
dtype: object

### Index
Provides the **indices** of the Dataframe, since each row is associated with its own index (1 to *n* by default).

In [8]:
ipl.index

RangeIndex(start=0, stop=950, step=1)

The above output, considering the `RangeIndex()` function suggests that an automatic index generation has ben used, taking numbers 1 to 950, starting at 0, and with 1 progressive index at each step.

### Columns
Provides the names of all the columns of the DataFrame.

In [9]:
movies.columns

Index(['title_x', 'imdb_id', 'poster_path', 'wiki_link', 'title_y',
       'original_title', 'is_adult', 'year_of_release', 'runtime', 'genres',
       'imdb_rating', 'imdb_votes', 'story', 'summary', 'tagline', 'actors',
       'wins_nominations', 'release_date'],
      dtype='object')

### Values

Going a little in depth with this one because I felt that this needs to be explained in a little bit detail.

The `.values` attribute of a **Pandas DataFrame** returns the **underlying data** as a **NumPy array** (or a similar array-like structure).

---

#### ✅ Syntax:

```python
df.values
```

---

#### 📌 What it returns:

* A **NumPy array** containing the **actual data** of the DataFrame **without any row or column labels**.
* The data types in the array will be upcasted if necessary (e.g., mixed types → object dtype). With this

---

#### 💻 Example:

```python
import pandas as pd

data = {
    'Name': ['Alice', 'Bob', 'Charlie'],
    'Age': [25, 30, 22],
    'Score': [88.5, 92.0, 79.5]
}

df = pd.DataFrame(data)
print(df)
```

**Output:**

```
      Name  Age  Score
0    Alice   25   88.5
1      Bob   30   92.0
2  Charlie   22   79.5
```

```python
print(df.values)
```

**Output:**

```python
array([['Alice', 25, 88.5],
       ['Bob', 30, 92.0],
       ['Charlie', 22, 79.5]], dtype=object)
```

---

#### 🧾 Notes:

* If all columns are of **numeric type**, the returned array will be of **numeric dtype** (e.g., `float64` or `int64`).
* If there are **mixed types** (like strings and numbers), the resulting array has `dtype=object`.
* `.values` **does not include** index or column labels — it is **pure data** only.

---

#### 🧪 Use Cases in Data Science:

* Useful for converting a DataFrame into a **NumPy array** for:

  * Mathematical operations
  * Feeding into machine learning models (e.g., Scikit-learn)
  * Custom numerical processing

---

#### ⚠️ Caution:

`.values` is not always preferred in modern pandas versions due to better alternatives:

* `df.to_numpy()` is recommended instead, because it handles corner cases more robustly.

```python
df.to_numpy()  # Safer, modern alternative
```

---

Would you like me to explain `.to_numpy()` and how it's different from `.values` next?


### Head and Tail
`head()` and `tail()` are demonstrated in the previous notes. These gives the top few values and the bottom most few values.

In [10]:
# head() demonstration
movies.head()

title_x    imdb_id  \
0              Uri: The Surgical Strike  tt8291224   
1                         Battalion 609  tt9472208   
2  The Accidental Prime Minister (film)  tt6986710   
3                       Why Cheat India  tt8108208   
4                       Evening Shadows  tt6028796   

                                         poster_path  \
0  https://upload.wikimedia.org/wikipedia/en/thum...   
1                                                NaN   
2  https://upload.wikimedia.org/wikipedia/en/thum...   
3  https://upload.wikimedia.org/wikipedia/en/thum...   
4                                                NaN   

                                           wiki_link  \
0  https://en.wikipedia.org/wiki/Uri:_The_Surgica...   
1        https://en.wikipedia.org/wiki/Battalion_609   
2  https://en.wikipedia.org/wiki/The_Accidental_P...   
3      https://en.wikipedia.org/wiki/Why_Cheat_India   
4      https://en.wikipedia.org/wiki/Evening_Shadows   

                         title_y                 original_title  is_adult  \
0       Uri: The Surgical Strike       Uri: The Surgical Strike         0   
1                  Battalion 609                  Battalion 609         0   
2  The Accidental Prime Minister  The Accidental Prime Minister         0   
3                Why Cheat India                Why Cheat India         0   
4                Evening Shadows                Evening Shadows         0   

   year_of_release runtime            genres  imdb_rating  imdb_votes  \
0             2019     138  Action|Drama|War          8.4       35112   
1             2019     131               War          4.1          73   
2             2019     112   Biography|Drama          6.1        5549   
3             2019     121       Crime|Drama          6.0        1891   
4             2018     102             Drama          7.3         280   

                                               story  \
0  Divided over five chapters  the film chronicle...   
1  The story revolves around a cricket match betw...   
2  Based on the memoir by Indian policy analyst S...   
3  The movie focuses on existing malpractices in ...   
4  While gay rights and marriage equality has bee...   

                                             summary tagline  \
0  Indian army special forces execute a covert op...     NaN   
1  The story of Battalion 609 revolves around a c...     NaN   
2  Explores Manmohan Singh's tenure as the Prime ...     NaN   
3  The movie focuses on existing malpractices in ...     NaN   
4  Under the 'Evening Shadows'  truth often plays...     NaN   

                                              actors        wins_nominations  \
0  Vicky Kaushal|Paresh Rawal|Mohit Raina|Yami Ga...                  4 wins   
1  Vicky Ahuja|Shoaib Ibrahim|Shrikant Kamat|Elen...                     NaN   
2  Anupam Kher|Akshaye Khanna|Aahana Kumra|Atul S...                     NaN   
3  Emraan Hashmi|Shreya Dhanwanthary|Snighdadeep ...                     NaN   
4  Mona Ambegaonkar|Ananth Narayan Mahadevan|Deva...  17 wins & 1 nomination   

              release_date  
0    11 January 2019 (USA)  
1  11 January 2019 (India)  
2    11 January 2019 (USA)  
3    18 January 2019 (USA)  
4  11 January 2019 (India)

In [11]:
# tail demonstration
ipl.tail()

ID        City        Date   Season MatchNumber  \
945  335986     Kolkata  2008-04-20  2007/08           4   
946  335985      Mumbai  2008-04-20  2007/08           5   
947  335984       Delhi  2008-04-19  2007/08           3   
948  335983  Chandigarh  2008-04-19  2007/08           2   
949  335982   Bangalore  2008-04-18  2007/08           1   

                           Team1                        Team2  \
945        Kolkata Knight Riders              Deccan Chargers   
946               Mumbai Indians  Royal Challengers Bangalore   
947             Delhi Daredevils             Rajasthan Royals   
948              Kings XI Punjab          Chennai Super Kings   
949  Royal Challengers Bangalore        Kolkata Knight Riders   

                                          Venue                   TossWinner  \
945                                Eden Gardens              Deccan Chargers   
946                            Wankhede Stadium               Mumbai Indians   
947                            Feroz Shah Kotla             Rajasthan Royals   
948  Punjab Cricket Association Stadium, Mohali          Chennai Super Kings   
949                       M Chinnaswamy Stadium  Royal Challengers Bangalore   

    TossDecision SuperOver                  WinningTeam    WonBy  Margin  \
945          bat         N        Kolkata Knight Riders  Wickets     5.0   
946          bat         N  Royal Challengers Bangalore  Wickets     5.0   
947          bat         N             Delhi Daredevils  Wickets     9.0   
948          bat         N          Chennai Super Kings     Runs    33.0   
949        field         N        Kolkata Knight Riders     Runs   140.0   

    method Player_of_Match                                       Team1Players  \
945    NaN       DJ Hussey  ['WP Saha', 'BB McCullum', 'RT Ponting', 'SC G...   
946    NaN      MV Boucher  ['L Ronchi', 'ST Jayasuriya', 'DJ Thornely', '...   
947    NaN     MF Maharoof  ['G Gambhir', 'V Sehwag', 'S Dhawan', 'MK Tiwa...   
948    NaN      MEK Hussey  ['K Goel', 'JR Hopes', 'KC Sangakkara', 'Yuvra...   
949    NaN     BB McCullum  ['R Dravid', 'W Jaffer', 'V Kohli', 'JH Kallis...   

                                          Team2Players    Umpire1  \
945  ['AC Gilchrist', 'Y Venugopal Rao', 'VVS Laxma...  BF Bowden   
946  ['S Chanderpaul', 'R Dravid', 'LRPL Taylor', '...   SJ Davis   
947  ['T Kohli', 'YK Pathan', 'SR Watson', 'M Kaif'...  Aleem Dar   
948  ['PA Patel', 'ML Hayden', 'MEK Hussey', 'MS Dh...  MR Benson   
949  ['SC Ganguly', 'BB McCullum', 'RT Ponting', 'D...  Asad Rauf   

            Umpire2  
945     K Hariharan  
946       DJ Harper  
947  GA Pratapkumar  
948      SL Shastri  
949     RE Koertzen

### Sample

The `sample()` function in pandas is used to **randomly select rows or columns** from a DataFrame. It is extremely useful in data science for tasks such as:

* Creating a **train-test split**
* Doing **random sampling** for exploratory data analysis (EDA)
* Testing pipelines with smaller data
* Bootstrapping methods

---

#### ✅ Syntax:

```python
df.sample(n=None, frac=None, replace=False, weights=None, random_state=None, axis=None)
```

---

#### 📌 Parameters:

| Parameter      | Description                                                     |
| -------------- | --------------------------------------------------------------- |
| `n`            | Number of items to return (rows or columns depending on `axis`) |
| `frac`         | Fraction of items to return (e.g. 0.5 for 50%)                  |
| `replace`      | If `True`, sample with replacement                              |
| `weights`      | Probability weights for sampling (can be a column or list)      |
| `random_state` | For reproducibility; sets the seed                              |
| `axis`         | 0 for rows (default), 1 for columns                             |

---

#### 🧪 Example 1: Random Row Sampling

```python
import pandas as pd

df = pd.DataFrame({
    'Name': ['Alice', 'Bob', 'Charlie', 'David', 'Eva'],
    'Score': [85, 92, 88, 76, 95]
})

# Sample 2 random rows
df.sample(n=2)
```

---

#### 🧪 Example 2: Sample 40% of the Data

```python
df.sample(frac=0.4)
```

---

#### 🧪 Example 3: Sampling with Replacement

`Replacement` enables the **repition** of chosen rows/columns. In case this is set as `False`, each row/col in the sample would be unique.
```python
df.sample(n=3, replace=True)
```

---

#### 🧪 Example 4: Sampling with a Fixed Random Seed (Reproducibility)

```python
df.sample(n=2, random_state=42)
```

This will always return the same rows on every run (good for experiments).

---

#### 🧪 Example 5: Column Sampling

```python
df.sample(n=1, axis=1)
```

Returns a random **column** instead of row.

---

#### 🎯 Use Cases in Data Science:

| Use Case         | Description                                       |
| ---------------- | ------------------------------------------------- |
| Train-Test Split | Randomly sample rows for training/testing         |
| Debugging        | Work on a small subset of the data                |
| Bootstrapping    | Create samples with replacement                   |
| Visual EDA       | Sample for quick plotting or insights             |
| Bias Detection   | Analyze if models are sensitive to random subsets |

Provided below are some of the demonstration done on the dataset that we previously imported.

In [12]:
# selecting 2 random samples
movies.sample(n=2)

title_x    imdb_id  \
529  Khoobsurat (2014 film)  tt3554418   
923      Kisse Pyaar Karoon  tt0438894   

                                           poster_path  \
529  https://upload.wikimedia.org/wikipedia/en/thum...   
923                                                NaN   

                                             wiki_link              title_y  \
529  https://en.wikipedia.org/wiki/Khoobsurat_(2014...           Khoobsurat   
923   https://en.wikipedia.org/wiki/Kisse_Pyaar_Karoon  Kisse Pyaar Karoon?   

          original_title  is_adult  year_of_release runtime  \
529           Khoobsurat         0             2014     130   
923  Kisse Pyaar Karoon?         0             2009     123   

                  genres  imdb_rating  imdb_votes  \
529       Comedy|Romance          6.4        6981   
923  Action|Comedy|Crime          3.3         120   

                                                 story  \
529  Khoobsurat is a quirky  modern romantic comedy...   
923  Mumbai-based collegian-slackers Sid(Arshad War...   

                                               summary tagline  \
529  A hopelessly romantic physiotherapist meets a ...     NaN   
923  Two men abduct a woman who wants to alienate t...     NaN   

                                                actors  \
529  Sonam Kapoor|Fawad Khan|Ratna Pathak Shah|Kiro...   
923  Arshad Warsi|Aashish Chaudhary|Yash Tonk|Udita...   

           wins_nominations              release_date  
529  2 wins & 2 nominations   19 September 2014 (USA)  
923                     NaN  27 February 2009 (India)

In [13]:
# random sampling 10% of the data
movies.sample(frac=0.1)

title_x    imdb_id  \
1004           Haal-e-Dil  tt1252488   
113               Phamous  tt8338746   
1614   Moksha (2001 film)  tt0301240   
1264           Koi Aap Sa  tt0488840   
799     Force (2011 film)  tt1992138   
...                   ...        ...   
952            Fox (film)  tt1324076   
982          Jodhaa Akbar  tt0449994   
1372         Phir Milenge  tt0422950   
1297  Sheesha (2005 film)  tt0445056   
316         Rustom (film)  tt5165344   

                                            poster_path  \
1004  https://upload.wikimedia.org/wikipedia/en/thum...   
113   https://upload.wikimedia.org/wikipedia/en/thum...   
1614  https://upload.wikimedia.org/wikipedia/en/6/6e...   
1264                                                NaN   
799   https://upload.wikimedia.org/wikipedia/en/thum...   
...                                                 ...   
952   https://upload.wikimedia.org/wikipedia/en/thum...   
982   https://upload.wikimedia.org/wikipedia/en/thum...   
1372  https://upload.wikimedia.org/wikipedia/en/thum...   
1297  https://upload.wikimedia.org/wikipedia/en/thum...   
316   https://upload.wikimedia.org/wikipedia/en/thum...   

                                              wiki_link  \
1004           https://en.wikipedia.org/wiki/Haal-e-Dil   
113               https://en.wikipedia.org/wiki/Phamous   
1614   https://en.wikipedia.org/wiki/Moksha_(2001_film)   
1264           https://en.wikipedia.org/wiki/Koi_Aap_Sa   
799     https://en.wikipedia.org/wiki/Force_(2011_film)   
...                                                 ...   
952            https://en.wikipedia.org/wiki/Fox_(film)   
982          https://en.wikipedia.org/wiki/Jodhaa_Akbar   
1372         https://en.wikipedia.org/wiki/Phir_Milenge   
1297  https://en.wikipedia.org/wiki/Sheesha_(2005_film)   
316         https://en.wikipedia.org/wiki/Rustom_(film)   

                                        title_y  \
1004                                 Haal-e-Dil   
113                                     Phamous   
1614                          Moksha: Salvation   
1264  Koi Aap Sa: But Lovers Have to Be Friends   
799                                       Force   
...                                         ...   
952                                         Fox   
982                                Jodhaa Akbar   
1372                               Phir Milenge   
1297                                    Sheesha   
316                                      Rustom   

                                 original_title  is_adult  year_of_release  \
1004                                 Haal-e-Dil         0             2008   
113                                     Phamous         0             2018   
1614                          Moksha: Salvation         0             2001   
1264  Koi Aap Sa: But Lovers Have to Be Friends         0             2005   
799                                       Force         0             2011   
...                                         ...       ...              ...   
952                                         Fox         0             2009   
982                                Jodhaa Akbar         0             2008   
1372                               Phir Milenge         0             2004   
1297                                    Sheesha         0             2005   
316                                      Rustom         0             2016   

     runtime                  genres  imdb_rating  imdb_votes  \
1004     124           Drama|Romance          3.4         224   
113      115      Comedy|Crime|Drama          3.7         207   
1614     150          Drama|Thriller          6.3         356   
1264     142    Comedy|Drama|Romance          5.4         194   
799      137         Action|Thriller          6.4        6497   
...      ...                     ...          ...         ...   
952      145     Crime|Drama|Mystery          5.2         512   
982      213    Action|Drama|History          7.6

In [14]:
ipl.sample(n=3, replace=True)

ID        City        Date   Season MatchNumber  \
831   419110     Chennai  2010-03-14  2009/10           5   
688   548319  Chandigarh  2012-04-12     2012          14   
246  1175363   Hyderabad  2019-03-29     2019           8   

                   Team1                Team2  \
831  Chennai Super Kings      Deccan Chargers   
688      Kings XI Punjab        Pune Warriors   
246     Rajasthan Royals  Sunrisers Hyderabad   

                                          Venue        TossWinner  \
831             MA Chidambaram Stadium, Chepauk   Deccan Chargers   
688  Punjab Cricket Association Stadium, Mohali   Kings XI Punjab   
246          Rajiv Gandhi International Stadium  Rajasthan Royals   

    TossDecision SuperOver          WinningTeam    WonBy  Margin method  \
831          bat         N      Deccan Chargers     Runs    31.0    NaN   
688        field         N      Kings XI Punjab  Wickets     7.0    NaN   
246          bat         N  Sunrisers Hyderabad  Wickets     5.0    NaN   

    Player_of_Match                                       Team1Players  \
831      WPUJC Vaas  ['M Vijay', 'ML Hayden', 'SK Raina', 'S Badrin...   
688  AD Mascarenhas  ['PC Valthaty', 'AC Gilchrist', 'SE Marsh', 'M...   
246     Rashid Khan  ['AM Rahane', 'JC Buttler', 'SV Samson', 'BA S...   

                                          Team2Players       Umpire1  \
831  ['AC Gilchrist', 'VVS Laxman', 'HH Gibbs', 'A ...   K Hariharan   
688  ['JD Ryder', 'SC Ganguly', 'MN Samuels', 'RV U...   VA Kulkarni   
246  ['DA Warner', 'JM Bairstow', 'KS Williamson', ...  BNJ Oxenford   

           Umpire2  
831      DJ Harper  
688    SK Tarapore  
246  C Shamshuddin

In [15]:
movies.sample(n=2, random_state=42)

title_x    imdb_id  \
669           Kahaani  tt1821480   
251  Monsoon Shootout  tt2198235   

                                           poster_path  \
669  https://upload.wikimedia.org/wikipedia/en/thum...   
251  https://upload.wikimedia.org/wikipedia/en/thum...   

                                          wiki_link           title_y  \
669           https://en.wikipedia.org/wiki/Kahaani           Kahaani   
251  https://en.wikipedia.org/wiki/Monsoon_Shootout  Monsoon Shootout   

       original_title  is_adult  year_of_release runtime              genres  \
669           Kahaani         0             2012     122    Mystery|Thriller   
251  Monsoon Shootout         0             2013      92  Action|Crime|Drama   

     imdb_rating  imdb_votes  \
669          8.1       53181   
251          6.5         801   

                                                 story  \
669  Kolkata is abuzz with the preparations for the...   
251  As the raging monsoon lashes Mumbai  the comme...   

                                               summary  \
669  A pregnant woman's search for her missing husb...   
251  As heavy rains lash Mumbai  a cop on his first...   

                                               tagline  \
669                                A mother of a story   
251  A rookie cop's moment of reckoning| to shoot o...   

                                                actors  \
669  Vidya Balan|Parambrata Chattopadhyay|Indraneil...   
251  Vijay Varma|Nawazuddin Siddiqui|Neeraj Kabi|Ge...   

             wins_nominations              release_date  
669  19 wins & 18 nominations      9 March 2012 (India)  
251     1 win & 7 nominations  15 December 2017 (India)

In [16]:
movies.sample(n=2, axis=1)

title_y  year_of_release
0          Uri: The Surgical Strike             2019
1                     Battalion 609             2019
2     The Accidental Prime Minister             2019
3                   Why Cheat India             2019
4                   Evening Shadows             2018
...                             ...              ...
1624          Tera Mera Saath Rahen             2001
1625           Yeh Zindagi Ka Safar             2001
1626                Sabse Bada Sukh             2018
1627                          Daaka             2019
1628                       Humsafar             2011

[1629 rows x 2 columns]

### Info
Provides a **High-Level information** about the DataFrame. This method prints information about a DataFrame including the **index dtype**, **columns**, **non-null values** and **memory usage**.

In [17]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1629 entries, 0 to 1628
Data columns (total 18 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   title_x           1629 non-null   object 
 1   imdb_id           1629 non-null   object 
 2   poster_path       1526 non-null   object 
 3   wiki_link         1629 non-null   object 
 4   title_y           1629 non-null   object 
 5   original_title    1629 non-null   object 
 6   is_adult          1629 non-null   int64  
 7   year_of_release   1629 non-null   int64  
 8   runtime           1629 non-null   object 
 9   genres            1629 non-null   object 
 10  imdb_rating       1629 non-null   float64
 11  imdb_votes        1629 non-null   int64  
 12  story             1609 non-null   object 
 13  summary           1629 non-null   object 
 14  tagline           557 non-null    object 
 15  actors            1624 non-null   object 
 16  wins_nominations  707 non-null    object 


### Describe
Describe is used to present some of the mathematical summary, further on which could be used for performing aggregate operations. Note that `describe()` identifies the columns that contain only numeric values and then provides their mathematical information.

In [18]:
movies.describe()

is_adult  year_of_release  imdb_rating     imdb_votes
count    1629.0      1629.000000  1629.000000    1629.000000
mean        0.0      2010.263966     5.557459    5384.263352
std         0.0         5.381542     1.567609   14552.103231
min         0.0      2001.000000     0.000000       0.000000
25%         0.0      2005.000000     4.400000     233.000000
50%         0.0      2011.000000     5.600000    1000.000000
75%         0.0      2015.000000     6.800000    4287.000000
max         0.0      2019.000000     9.400000  310481.000000

### Isnull
Reconstrcuts a **Boolean** DataFrame over the provided DataFrame, placing the values as `True` where there are null values and `False` where there aren't.

In [19]:
movies.isnull()

title_x  imdb_id  poster_path  wiki_link  title_y  original_title  \
0       False    False        False      False    False           False   
1       False    False         True      False    False           False   
2       False    False        False      False    False           False   
3       False    False        False      False    False           False   
4       False    False         True      False    False           False   
...       ...      ...          ...        ...      ...             ...   
1624    False    False        False      False    False           False   
1625    False    False        False      False    False           False   
1626    False    False         True      False    False           False   
1627    False    False        False      False    False           False   
1628    False    False        False      False    False           False   

      is_adult  year_of_release  runtime  genres  imdb_rating  imdb_votes  \
0        False            False    False   False        False       False   
1        False            False    False   False        False       False   
2        False            False    False   False        False       False   
3        False            False    False   False        False       False   
4        False            False    False   False        False       False   
...        ...              ...      ...     ...          ...         ...   
1624     False            False    False   False        False       False   
1625     False            False    False   False        False       False   
1626     False            False    False   False        False       False   
1627     False            False    False   False        False       False   
1628     False            False    False   False        False       False   

      story  summary  tagline  actors  wins_nominations  release_date  
0     False    False     True   False             False         False  
1     False    False     True   False              True         False  
2     False    False     True   False              True         False  
3     False    False     True   False              True         False  
4     False    False     True   False             False         False  
...     ...      ...      ...     ...               ...           ...  
1624  False    False     True   False              True         False  
1625  False    False     True   False              True         False  
1626  False    False     True   False              True          True  
1627  False    False     True   False              True         False  
1628  False    False     True   False              True         False  

[1629 rows x 18 columns]

However, at times, we may need to see which of the column has *null* values, and how much. For this, we can sum up the count of the `True` values.

In [20]:
movies.isnull().sum()

title_x                0
imdb_id                0
poster_path          103
wiki_link              0
title_y                0
original_title         0
is_adult               0
year_of_release        0
runtime                0
genres                 0
imdb_rating            0
imdb_votes             0
story                 20
summary                0
tagline             1072
actors                 5
wins_nominations     922
release_date         107
dtype: int64

### Duplicated
Displays whether there are **duplicate data** present in the data or not. Since duplicated data are always *pointless*, they serve no purpose and shall be removed upon identification.

In [21]:
movies.duplicated().sum()

np.int64(0)

In [22]:
ipl.duplicated().sum()

np.int64(0)

In [23]:
print(stud.duplicated())
stud.duplicated().sum()

0    False
1    False
2    False
3    False
4     True
dtype: bool


np.int64(1)

### Rename
`rename()` is used to temporatily rename the column names of a DataFrame. However, the changes made using this method will be temporary, unless they are stored in a variable.

In [24]:
stud.rename(columns={'marks': 'percent', 'package': 'lpa'})

iq  percent  pkg
0  100       80   10
1   90       70    7
2  120      100   14
3   80        5    2
4  100       80   10

However upon rechecking the original `stud` variable, we will again get the original column names.

In [25]:
stud

iq  marks  pkg
0  100     80   10
1   90     70    7
2  120    100   14
3   80      5    2
4  100     80   10

Note that *parameter* `inplace` can be set to `True` to make the changes permanent, in the original variable.

## Mathematical Methods
In this section, we will be discussing some of the mathematical functions available over DataFrames. A lot of these would be inspired from those of series, since a DataFrame is a collaboration of mutiple Series.

### Sum
Applies `sum()` over all the columns of the DataFrame. Note that in case the **dtype** of the Series is *string* or *float*, in such a case the strings will be **concatenated**.

In most of the cases, it is not logical to perform **sum** over all the columns of the DataFrames. Additionally, `sum()` can also be performed on the rows of the DataFrame by setting the `axis` param as `1`.

In [26]:
stud.sum()

iq       490
marks    335
pkg       43
dtype: int64

In [27]:
# for row wise sum, although this isn't used much
stud.sum(axis=1)

0    190
1    167
2    234
3     87
4    190
dtype: int64

Similarly, we have other functionalities such as `min()`, `max()`, `mean()`, `median()`, `std()`, and `var()`.

## Selecting *specific* **Columns** from a DataFrame

In a lot of scenarios, we may be required to select specific Columns from the Dataframe, rather than selecting the whole dataframe. This can be done using the following ways:

In [28]:
# to display the column 'Venue' of DF ipl
ipl['Venue']

0                Narendra Modi Stadium, Ahmedabad
1                Narendra Modi Stadium, Ahmedabad
2                           Eden Gardens, Kolkata
3                           Eden Gardens, Kolkata
4                        Wankhede Stadium, Mumbai
                          ...                    
945                                  Eden Gardens
946                              Wankhede Stadium
947                              Feroz Shah Kotla
948    Punjab Cricket Association Stadium, Mohali
949                         M Chinnaswamy Stadium
Name: Venue, Length: 950, dtype: object

In the above code, we are selecting a single column and so it will be of type **Series**, whereas multiple columns will result in a data of type **DataFrame**.

Moreover, we will now try to extract multiple columns from a DF using a method similar to *Slicing*. The order of columns will be decided by the placement of the columns in the code responsible for it.

In [29]:
movies[['title_x', 'year_of_release', 'actors']]

title_x  year_of_release  \
0                 Uri: The Surgical Strike             2019   
1                            Battalion 609             2019   
2     The Accidental Prime Minister (film)             2019   
3                          Why Cheat India             2019   
4                          Evening Shadows             2018   
...                                    ...              ...   
1624                 Tera Mera Saath Rahen             2001   
1625                  Yeh Zindagi Ka Safar             2001   
1626                       Sabse Bada Sukh             2018   
1627                                 Daaka             2019   
1628                              Humsafar             2011   

                                                 actors  
0     Vicky Kaushal|Paresh Rawal|Mohit Raina|Yami Ga...  
1     Vicky Ahuja|Shoaib Ibrahim|Shrikant Kamat|Elen...  
2     Anupam Kher|Akshaye Khanna|Aahana Kumra|Atul S...  
3     Emraan Hashmi|Shreya Dhanwanthary|Snighdadeep ...  
4     Mona Ambegaonkar|Ananth Narayan Mahadevan|Deva...  
...                                                 ...  
1624  Ajay Devgn|Sonali Bendre|Namrata Shirodkar|Pre...  
1625  Ameesha Patel|Jimmy Sheirgill|Nafisa Ali|Gulsh...  
1626  Vijay Arora|Asrani|Rajni Bala|Kumud Damle|Utpa...  
1627                          Gippy Grewal|Zareen Khan|  
1628                                        Fawad Khan|  

[1629 rows x 3 columns]

In [30]:
print(type(movies[['title_x', 'year_of_release', 'actors']]), type(ipl['Venue']))

<class 'pandas.core.frame.DataFrame'> <class 'pandas.core.series.Series'>


## Selecting *specific* **Rows** from a DataFrame
In the similar way as above, we can also select specific rows from a DataFrame.

For this, we must first learn about *two* methods:
1. `iloc`: searches using **index positions**
2. `loc`: searches using **index labels**

Here is a basic understanding of `iloc` and `loc`, and then we will move to the usecases on our own dataframe.

## 📍 `iloc` vs `loc` in Pandas

`iloc` and `loc` are two of the most important **data selection methods** in pandas. They help in accessing rows and columns of a DataFrame in different ways:

---

### 🔹 `iloc` (Integer Location Based Indexing)

* Stands for **integer-location based indexing**
* Access rows/columns by **position**
* Uses **integer indexes**
* **Exclusive of the end index** in slicing (like Python lists)

#### 🔸 Syntax:

```python
df.iloc[<row_index>, <column_index>]
```

#### ✅ Examples:

```python
df.iloc[0]            # First row
df.iloc[0, 1]         # Value at 1st row and 2nd column
df.iloc[2:5]          # Rows from index 2 to 4
df.iloc[:, 0]         # All rows, 1st column
```

---

### 🔹 `loc` (Label Based Indexing)

* Stands for **label-based indexing**
* Access rows/columns by **labels/names**
* Uses **row/column labels**
* **Inclusive of the end label** in slicing

#### 🔸 Syntax:

```python
df.loc[<row_label>, <column_label>]
```

#### ✅ Examples:

```python
df.loc[0]                    # Row with index label 0
df.loc[0, 'Name']            # Value from row label 0 and column 'Name'
df.loc[1:3, ['Name', 'Score']]  # Multiple row and column labels
df.loc[:, 'Score']           # All rows, 'Score' column
```

---

### 🎯 Use Cases

| Use Case                                                | `iloc`                    | `loc`                      |
| ------------------------------------------------------- | ------------------------- | -------------------------- |
| Working with positions (e.g. nth row/column)            | ✅ Best suited             | ❌ Not applicable           |
| Accessing by label (e.g. 'Age', 'Name')                 | ❌ Not allowed             | ✅ Required                 |
| Iterating row-wise/column-wise numerically              | ✅                         | ❌                          |
| Filtering with logical conditions (on index or columns) | ❌ Only works with numeric | ✅ Fully supported          |
| Selecting a column by name                              | ❌ Use `iloc[:, index]`    | ✅ Use `loc[:, 'col_name']` |
| Slicing including last element                          | ❌ Exclusive end           | ✅ Inclusive end            |

---

### ⚠️ Quick Comparison

| Feature         | `iloc`          | `loc`                       |
| --------------- | --------------- | --------------------------- |
| Index type      | Integer         | Label                       |
| Syntax          | `df.iloc[2, 1]` | `df.loc[2, 'Name']`         |
| Slice behavior  | Exclusive end   | Inclusive end               |
| Error if label? | Yes             | No                          |
| Error if int?   | No              | Yes (unless int is a label) |

---

### ✅ **Core Difference Between `iloc` and `loc`**

| Feature                    | `iloc`                                     | `loc`                                                  |
| -------------------------- | ------------------------------------------ | ------------------------------------------------------ |
| **Based on**               | **Integer positions** (0-based)            | **Labels** or names of rows/columns                    |
| **Indexing**               | Works like Python lists                    | Works like dictionary keys                             |
| **Use case**               | When you know the position                 | When you know the label                                |
| **Slicing**                | End index is **exclusive** (`[start:end)`) | End index is **inclusive** (`[start:end]`)             |
| **Error on label**         | ❌ Cannot use string labels                 | ✅ Can use string labels                                |
| **Error on integer index** | ✅ Accepts any valid integer                | ❌ Fails if label does not exist or is not integer type |

---

### 🎯 In One Line:

> **`iloc`** is **position-based**,
> **`loc`** is **label-based**.

---

### 🔍 Example:

```python
import pandas as pd

df = pd.DataFrame({
    "Name": ["Alice", "Bob", "Charlie"],
    "Age": [25, 30, 35]
}, index=["a", "b", "c"])
```

#### Using `iloc`:

```python
df.iloc[0]        # First row (index 'a')
df.iloc[1, 1]     # Value 30 (row 2, column 2)
```

#### Using `loc`:

```python
df.loc["a"]       # Row with label 'a'
df.loc["b", "Age"] # Value 30 (label 'b', column 'Age')
```

### `iloc`

In [8]:
import pandas as pd

df = pd.DataFrame({
    "Name": ["Alice", "Bob", "Charlie"],
    "Age": [25, 30, 35]
}, index=["a", "b", "c"])

df.iloc[0]

Name    Alice
Age        25
Name: a, dtype: object

In [31]:
# single row
print(movies.iloc[2])
type(movies.iloc[2])

title_x                          The Accidental Prime Minister (film)
imdb_id                                                     tt6986710
poster_path         https://upload.wikimedia.org/wikipedia/en/thum...
wiki_link           https://en.wikipedia.org/wiki/The_Accidental_P...
title_y                                 The Accidental Prime Minister
original_title                          The Accidental Prime Minister
is_adult                                                            0
year_of_release                                                  2019
runtime                                                           112
genres                                                Biography|Drama
imdb_rating                                                       6.1
imdb_votes                                                       5549
story               Based on the memoir by Indian policy analyst S...
summary             Explores Manmohan Singh's tenure as the Prime ...
tagline             

pandas.core.series.Series

In [32]:
# multiple rows
print(type(movies.iloc[0:5]))
movies.iloc[0:5]

<class 'pandas.core.frame.DataFrame'>


title_x    imdb_id  \
0              Uri: The Surgical Strike  tt8291224   
1                         Battalion 609  tt9472208   
2  The Accidental Prime Minister (film)  tt6986710   
3                       Why Cheat India  tt8108208   
4                       Evening Shadows  tt6028796   

                                         poster_path  \
0  https://upload.wikimedia.org/wikipedia/en/thum...   
1                                                NaN   
2  https://upload.wikimedia.org/wikipedia/en/thum...   
3  https://upload.wikimedia.org/wikipedia/en/thum...   
4                                                NaN   

                                           wiki_link  \
0  https://en.wikipedia.org/wiki/Uri:_The_Surgica...   
1        https://en.wikipedia.org/wiki/Battalion_609   
2  https://en.wikipedia.org/wiki/The_Accidental_P...   
3      https://en.wikipedia.org/wiki/Why_Cheat_India   
4      https://en.wikipedia.org/wiki/Evening_Shadows   

                         title_y                 original_title  is_adult  \
0       Uri: The Surgical Strike       Uri: The Surgical Strike         0   
1                  Battalion 609                  Battalion 609         0   
2  The Accidental Prime Minister  The Accidental Prime Minister         0   
3                Why Cheat India                Why Cheat India         0   
4                Evening Shadows                Evening Shadows         0   

   year_of_release runtime            genres  imdb_rating  imdb_votes  \
0             2019     138  Action|Drama|War          8.4       35112   
1             2019     131               War          4.1          73   
2             2019     112   Biography|Drama          6.1        5549   
3             2019     121       Crime|Drama          6.0        1891   
4             2018     102             Drama          7.3         280   

                                               story  \
0  Divided over five chapters  the film chronicle...   
1  The story revolves around a cricket match betw...   
2  Based on the memoir by Indian policy analyst S...   
3  The movie focuses on existing malpractices in ...   
4  While gay rights and marriage equality has bee...   

                                             summary tagline  \
0  Indian army special forces execute a covert op...     NaN   
1  The story of Battalion 609 revolves around a c...     NaN   
2  Explores Manmohan Singh's tenure as the Prime ...     NaN   
3  The movie focuses on existing malpractices in ...     NaN   
4  Under the 'Evening Shadows'  truth often plays...     NaN   

                                              actors        wins_nominations  \
0  Vicky Kaushal|Paresh Rawal|Mohit Raina|Yami Ga...                  4 wins   
1  Vicky Ahuja|Shoaib Ibrahim|Shrikant Kamat|Elen...                     NaN   
2  Anupam Kher|Akshaye Khanna|Aahana Kumra|Atul S...                     NaN   
3  Emraan Hashmi|Shreya Dhanwanthary|Snighdadeep ...                     NaN   
4  Mona Ambegaonkar|Ananth Narayan Mahadevan|Deva...  17 wins & 1 nomination   

              release_date  
0    11 January 2019 (USA)  
1  11 January 2019 (India)  
2    11 January 2019 (USA)  
3    18 January 2019 (USA)  
4  11 January 2019 (India)

Now, fancy indexing can also be used to fetch any number of rows.

In [33]:
movies.iloc[[0, 4, 5]]

title_x    imdb_id  \
0  Uri: The Surgical Strike  tt8291224   
4           Evening Shadows  tt6028796   
5               Soni (film)  tt6078866   

                                         poster_path  \
0  https://upload.wikimedia.org/wikipedia/en/thum...   
4                                                NaN   
5  https://upload.wikimedia.org/wikipedia/en/thum...   

                                           wiki_link  \
0  https://en.wikipedia.org/wiki/Uri:_The_Surgica...   
4      https://en.wikipedia.org/wiki/Evening_Shadows   
5          https://en.wikipedia.org/wiki/Soni_(film)   

                    title_y            original_title  is_adult  \
0  Uri: The Surgical Strike  Uri: The Surgical Strike         0   
4           Evening Shadows           Evening Shadows         0   
5                      Soni                      Soni         0   

   year_of_release runtime            genres  imdb_rating  imdb_votes  \
0             2019     138  Action|Drama|War          8.4       35112   
4             2018     102             Drama          7.3         280   
5             2018      97             Drama          7.2        1595   

                                               story  \
0  Divided over five chapters  the film chronicle...   
4  While gay rights and marriage equality has bee...   
5  Soni  a young policewoman in Delhi  and her su...   

                                             summary tagline  \
0  Indian army special forces execute a covert op...     NaN   
4  Under the 'Evening Shadows'  truth often plays...     NaN   
5  While fighting crimes against women in Delhi  ...     NaN   

                                              actors        wins_nominations  \
0  Vicky Kaushal|Paresh Rawal|Mohit Raina|Yami Ga...                  4 wins   
4  Mona Ambegaonkar|Ananth Narayan Mahadevan|Deva...  17 wins & 1 nomination   
5  Geetika Vidya Ohlyan|Saloni Batra|Vikas Shukla...  3 wins & 5 nominations   

              release_date  
0    11 January 2019 (USA)  
4  11 January 2019 (India)  
5    18 January 2019 (USA)

### `loc`

In [34]:
students.set_index('name', inplace=True)

In [35]:
students.loc['nool']

iq       100
marks     80
pkg       10
Name: nool, dtype: int64

In [36]:
# fancy indexing using loc
students.loc[['nool', 'rupaid', 'amnit']]

iq  marks  pkg
name                   
nool    100     80   10
rupaid  120    100   14
amnit     0      0    0

## Selecting both Rows and Columns

Suppose we want to fetch the data of first 3 movies, but only require the information from the first three columns:

In [37]:
movies.iloc[0:3, 0:3]

title_x    imdb_id  \
0              Uri: The Surgical Strike  tt8291224   
1                         Battalion 609  tt9472208   
2  The Accidental Prime Minister (film)  tt6986710   

                                         poster_path  
0  https://upload.wikimedia.org/wikipedia/en/thum...  
1                                                NaN  
2  https://upload.wikimedia.org/wikipedia/en/thum...

In [38]:
# similar thing using loc
movies.loc[0:3, 'title_x': 'poster_path']

title_x    imdb_id  \
0              Uri: The Surgical Strike  tt8291224   
1                         Battalion 609  tt9472208   
2  The Accidental Prime Minister (film)  tt6986710   
3                       Why Cheat India  tt8108208   

                                         poster_path  
0  https://upload.wikimedia.org/wikipedia/en/thum...  
1                                                NaN  
2  https://upload.wikimedia.org/wikipedia/en/thum...  
3  https://upload.wikimedia.org/wikipedia/en/thum...

## Filtering a DataFrame
Here, we will discuss about selecting some specific rows and columns based upon the requirements.

In [39]:
ipl.head(2)

ID       City        Date Season  MatchNumber  \
0  1312200  Ahmedabad  2022-05-29   2022        Final   
1  1312199  Ahmedabad  2022-05-27   2022  Qualifier 2   

                         Team1             Team2  \
0             Rajasthan Royals    Gujarat Titans   
1  Royal Challengers Bangalore  Rajasthan Royals   

                              Venue        TossWinner TossDecision SuperOver  \
0  Narendra Modi Stadium, Ahmedabad  Rajasthan Royals          bat         N   
1  Narendra Modi Stadium, Ahmedabad  Rajasthan Royals        field         N   

        WinningTeam    WonBy  Margin method Player_of_Match  \
0    Gujarat Titans  Wickets     7.0    NaN       HH Pandya   
1  Rajasthan Royals  Wickets     7.0    NaN      JC Buttler   

                                        Team1Players  \
0  ['YBK Jaiswal', 'JC Buttler', 'SV Samson', 'D ...   
1  ['V Kohli', 'F du Plessis', 'RM Patidar', 'GJ ...   

                                        Team2Players      Umpire1      Umpire2  
0  ['WP Saha', 'Shubman Gill', 'MS Wade', 'HH Pan...  CB Gaffaney  Nitin Menon  
1  ['YBK Jaiswal', 'JC Buttler', 'SV Samson', 'D ...  CB Gaffaney  Nitin Menon

### Task: Find all the final winners

In [40]:
# find all the final winners
mask = ipl['MatchNumber'] == 'Final'
finalMatch_df = ipl[mask]
finalMatch_df[['Season', 'Team1', 'Team2', 'WinningTeam']]

Season                        Team1                        Team2  \
0       2022             Rajasthan Royals               Gujarat Titans   
74      2021          Chennai Super Kings        Kolkata Knight Riders   
134  2020/21               Delhi Capitals               Mumbai Indians   
194     2019               Mumbai Indians          Chennai Super Kings   
254     2018          Sunrisers Hyderabad          Chennai Super Kings   
314     2017               Mumbai Indians       Rising Pune Supergiant   
373     2016  Royal Challengers Bangalore          Sunrisers Hyderabad   
433     2015               Mumbai Indians          Chennai Super Kings   
492     2014        Kolkata Knight Riders              Kings XI Punjab   
552     2013          Chennai Super Kings               Mumbai Indians   
628     2012        Kolkata Knight Riders          Chennai Super Kings   
702     2011          Chennai Super Kings  Royal Challengers Bangalore   
775  2009/10          Chennai Super Kings               Mumbai Indians   
835     2009  Royal Challengers Bangalore              Deccan Chargers   
892  2007/08          Chennai Super Kings             Rajasthan Royals   

               WinningTeam  
0           Gujarat Titans  
74     Chennai Super Kings  
134         Mumbai Indians  
194         Mumbai Indians  
254    Chennai Super Kings  
314         Mumbai Indians  
373    Sunrisers Hyderabad  
433         Mumbai Indians  
492  Kolkata Knight Riders  
552         Mumbai Indians  
628  Kolkata Knight Riders  
702    Chennai Super Kings  
775    Chennai Super Kings  
835        Deccan Chargers  
892       Rajasthan Royals

Or, we can simply sum up the above code in the following manner:

In [41]:
ipl[ipl['MatchNumber'] == 'Final'][['Season', 'Team1', 'Team2', 'WinningTeam']]

Season                        Team1                        Team2  \
0       2022             Rajasthan Royals               Gujarat Titans   
74      2021          Chennai Super Kings        Kolkata Knight Riders   
134  2020/21               Delhi Capitals               Mumbai Indians   
194     2019               Mumbai Indians          Chennai Super Kings   
254     2018          Sunrisers Hyderabad          Chennai Super Kings   
314     2017               Mumbai Indians       Rising Pune Supergiant   
373     2016  Royal Challengers Bangalore          Sunrisers Hyderabad   
433     2015               Mumbai Indians          Chennai Super Kings   
492     2014        Kolkata Knight Riders              Kings XI Punjab   
552     2013          Chennai Super Kings               Mumbai Indians   
628     2012        Kolkata Knight Riders          Chennai Super Kings   
702     2011          Chennai Super Kings  Royal Challengers Bangalore   
775  2009/10          Chennai Super Kings               Mumbai Indians   
835     2009  Royal Challengers Bangalore              Deccan Chargers   
892  2007/08          Chennai Super Kings             Rajasthan Royals   

               WinningTeam  
0           Gujarat Titans  
74     Chennai Super Kings  
134         Mumbai Indians  
194         Mumbai Indians  
254    Chennai Super Kings  
314         Mumbai Indians  
373    Sunrisers Hyderabad  
433         Mumbai Indians  
492  Kolkata Knight Riders  
552         Mumbai Indians  
628  Kolkata Knight Riders  
702    Chennai Super Kings  
775    Chennai Super Kings  
835        Deccan Chargers  
892       Rajasthan Royals

### Task: How many super over finishes have occured?

In [42]:
# strategy 1
(ipl['SuperOver'] == 'Y').sum()

np.int64(14)

In [43]:
# strategy 2
ipl[ipl['SuperOver'] == 'Y'].shape[0]

14

### Task: How many matches have "Chennai Super Kings" won in **Kolkata**?

In [58]:
ipl[(ipl['City'] == 'Kolkata') & (ipl['WinningTeam'] == 'Chennai Super Kings')].shape[0]

5

### Task: The percentage of **Toss Winner** also being the **Match Winner**

In [70]:
times_succeeded = ipl[ipl['TossWinner'] == ipl['WinningTeam']].shape[0]
total_matches = ipl.shape[0]
percentage = (times_succeeded / total_matches) * 100
print(f"The percentage of Toss Winner also being the Winning Team is {round(percentage, 2)}%")
print(times_succeeded, total_matches)

The percentage of Toss Winner also being the Winning Team is 51.47%
489 950


### Task: Movies with rating higher than 8 and votes > 10000

In [80]:
movies.head(2)

title_x    imdb_id  \
0  Uri: The Surgical Strike  tt8291224   
1             Battalion 609  tt9472208   

                                         poster_path  \
0  https://upload.wikimedia.org/wikipedia/en/thum...   
1                                                NaN   

                                           wiki_link  \
0  https://en.wikipedia.org/wiki/Uri:_The_Surgica...   
1        https://en.wikipedia.org/wiki/Battalion_609   

                    title_y            original_title  is_adult  \
0  Uri: The Surgical Strike  Uri: The Surgical Strike         0   
1             Battalion 609             Battalion 609         0   

   year_of_release runtime            genres  imdb_rating  imdb_votes  \
0             2019     138  Action|Drama|War          8.4       35112   
1             2019     131               War          4.1          73   

                                               story  \
0  Divided over five chapters  the film chronicle...   
1  The story revolves around a cricket match betw...   

                                             summary tagline  \
0  Indian army special forces execute a covert op...     NaN   
1  The story of Battalion 609 revolves around a c...     NaN   

                                              actors wins_nominations  \
0  Vicky Kaushal|Paresh Rawal|Mohit Raina|Yami Ga...           4 wins   
1  Vicky Ahuja|Shoaib Ibrahim|Shrikant Kamat|Elen...              NaN   

              release_date  
0    11 January 2019 (USA)  
1  11 January 2019 (India)

In [83]:
movies[(movies['imdb_rating'] > 8) & (movies['imdb_votes'] > 10000)].shape[0]

43

### Task: Action movies (but not mandatorily only Action) with rating higher than 7.5

For this task, the way to do this is not a direct one. Since a movie can have **Genres**, we will have to split all the Genres, forming a list of Genres.

In [88]:
mask1 = movies['genres'].str.split('|').apply(lambda x: 'Action' in x)
mask2 = movies['imdb_rating'] > 7.5

Here is a brief explanation of the above code. 
- The 'genres' must be converted to type `str` for split to work on it.
- Then we create a list out of it using `split()`
- We then **apply** our single line logic, via `lambda` function, using the *membership operator* `in` to iterate over the list elements and check if 'Action' exists within or not.
- These filters are stored as variables `mask1` and `mask2`, which is then used along with `&` for picking specific columns out of the DF **movies** in the following code cell.

In [91]:
movies[(mask1) & (mask2)].shape[0]

33

In [94]:
# or we could also create the mask using this strategy
mask1 = movies['genres'].str.contains('Action')
mask2 = movies['imdb_rating'] > 7.5
movies[(mask1) & (mask2)].shape[0]

33

### Task: Write a function that receives the names of 2 teams from the user and returns their track record.

In [95]:
ipl.head(2)

ID       City        Date Season  MatchNumber  \
0  1312200  Ahmedabad  2022-05-29   2022        Final   
1  1312199  Ahmedabad  2022-05-27   2022  Qualifier 2   

                         Team1             Team2  \
0             Rajasthan Royals    Gujarat Titans   
1  Royal Challengers Bangalore  Rajasthan Royals   

                              Venue        TossWinner TossDecision SuperOver  \
0  Narendra Modi Stadium, Ahmedabad  Rajasthan Royals          bat         N   
1  Narendra Modi Stadium, Ahmedabad  Rajasthan Royals        field         N   

        WinningTeam    WonBy  Margin method Player_of_Match  \
0    Gujarat Titans  Wickets     7.0    NaN       HH Pandya   
1  Rajasthan Royals  Wickets     7.0    NaN      JC Buttler   

                                        Team1Players  \
0  ['YBK Jaiswal', 'JC Buttler', 'SV Samson', 'D ...   
1  ['V Kohli', 'F du Plessis', 'RM Patidar', 'GJ ...   

                                        Team2Players      Umpire1      Umpire2  
0  ['WP Saha', 'Shubman Gill', 'MS Wade', 'HH Pan...  CB Gaffaney  Nitin Menon  
1  ['YBK Jaiswal', 'JC Buttler', 'SV Samson', 'D ...  CB Gaffaney  Nitin Menon

In [111]:
teams_list = ipl.apply(lambda row: [row['Team1'], row['Team2']], axis=1)
teams_list

0                     [Rajasthan Royals, Gujarat Titans]
1        [Royal Challengers Bangalore, Rajasthan Royals]
2      [Royal Challengers Bangalore, Lucknow Super Gi...
3                     [Rajasthan Royals, Gujarat Titans]
4                    [Sunrisers Hyderabad, Punjab Kings]
                             ...                        
945             [Kolkata Knight Riders, Deccan Chargers]
946        [Mumbai Indians, Royal Challengers Bangalore]
947                 [Delhi Daredevils, Rajasthan Royals]
948               [Kings XI Punjab, Chennai Super Kings]
949    [Royal Challengers Bangalore, Kolkata Knight R...
Length: 950, dtype: object

In [110]:
teams_list = ipl.apply(lambda row: row["Team1"] + ' ' + row['Team2'], axis=1)
teams_list

0                        Rajasthan Royals Gujarat Titans
1           Royal Challengers Bangalore Rajasthan Royals
2       Royal Challengers Bangalore Lucknow Super Giants
3                        Rajasthan Royals Gujarat Titans
4                       Sunrisers Hyderabad Punjab Kings
                             ...                        
945                Kolkata Knight Riders Deccan Chargers
946           Mumbai Indians Royal Challengers Bangalore
947                    Delhi Daredevils Rajasthan Royals
948                  Kings XI Punjab Chennai Super Kings
949    Royal Challengers Bangalore Kolkata Knight Riders
Length: 950, dtype: object

In [ ]:
def trackRecord(team1, team2, ipl_df):
    teams_list = ipl_df.apply(lambda row: [row['Team1'], row['Team2']], axis=1)
    ipl_df

## Adding New Columns
We can add new columns to our Dataframe, either by **creating** a completely new one or by using an old **existing series**.

In [112]:
movies.head(2)

title_x    imdb_id  \
0  Uri: The Surgical Strike  tt8291224   
1             Battalion 609  tt9472208   

                                         poster_path  \
0  https://upload.wikimedia.org/wikipedia/en/thum...   
1                                                NaN   

                                           wiki_link  \
0  https://en.wikipedia.org/wiki/Uri:_The_Surgica...   
1        https://en.wikipedia.org/wiki/Battalion_609   

                    title_y            original_title  is_adult  \
0  Uri: The Surgical Strike  Uri: The Surgical Strike         0   
1             Battalion 609             Battalion 609         0   

   year_of_release runtime            genres  imdb_rating  imdb_votes  \
0             2019     138  Action|Drama|War          8.4       35112   
1             2019     131               War          4.1          73   

                                               story  \
0  Divided over five chapters  the film chronicle...   
1  The story revolves around a cricket match betw...   

                                             summary tagline  \
0  Indian army special forces execute a covert op...     NaN   
1  The story of Battalion 609 revolves around a c...     NaN   

                                              actors wins_nominations  \
0  Vicky Kaushal|Paresh Rawal|Mohit Raina|Yami Ga...           4 wins   
1  Vicky Ahuja|Shoaib Ibrahim|Shrikant Kamat|Elen...              NaN   

              release_date  
0    11 January 2019 (USA)  
1  11 January 2019 (India)

In [ ]:
# Creating a new column, containing the same value for each row:
movies['Country'] = 'India'

In [7]:
# creating one from an existing one
notNullMovies = movies.dropna()
notNullMovies['lead actor'] = notNullMovies['actors'].str.split('|').apply(lambda x: x[0])
notNullMovies.head()

title_x     imdb_id  \
11          Gully Boy   tt2395469   
34      Yeh Hai India   tt5525846   
37  Article 15 (film)  tt10324144   
87            Aiyaary   tt6774212   
96   Raid (2018 film)   tt7363076   

                                          poster_path  \
11  https://upload.wikimedia.org/wikipedia/en/thum...   
34  https://upload.wikimedia.org/wikipedia/en/thum...   
37  https://upload.wikimedia.org/wikipedia/en/thum...   
87  https://upload.wikimedia.org/wikipedia/en/thum...   
96  https://upload.wikimedia.org/wikipedia/en/thum...   

                                          wiki_link        title_y  \
11          https://en.wikipedia.org/wiki/Gully_Boy      Gully Boy   
34      https://en.wikipedia.org/wiki/Yeh_Hai_India  Yeh Hai India   
37  https://en.wikipedia.org/wiki/Article_15_(film)     Article 15   
87            https://en.wikipedia.org/wiki/Aiyaary        Aiyaary   
96   https://en.wikipedia.org/wiki/Raid_(2018_film)           Raid   

   original_title  is_adult  year_of_release runtime                  genres  \
11      Gully Boy         0             2019     153             Drama|Music   
34  Yeh Hai India         0             2017     128  Action|Adventure|Drama   
37     Article 15         0             2019     130             Crime|Drama   
87        Aiyaary         0             2018     157         Action|Thriller   
96           Raid         0             2018     122      Action|Crime|Drama   

    imdb_rating  imdb_votes  \
11          8.2       22440   
34          5.7         169   
37          8.3       13417   
87          5.2        3538   
96          7.4       13159   

                                                story  \
11  Gully Boy is a film about a 22-year-old boy "M...   
34  Yeh Hai India  follows the story of a 25 years...   
37  In the rural heartlands of India  an upright p...   
87  General Gurinder Singh comes with a proposal t...   
96  Set in the 80s in Uttar Pradesh  India  Raid i...   

                                              summary  \
11  A coming-of-age story based on the lives of st...   
34  Yeh Hai India  follows the story of a 25 years...   
37  In the rural heartlands of India  an upright p...   
87  After finding out about an illegal arms deal  ...   
96  A fearless income tax officer raids the mansio...   

                                   tagline  \
11                       Apna Time Aayega!   
34                 A Film for Every Indian   
37  Farq Bahut Kar Liya| Ab Farq Laayenge.   
87                   The Ultimate Trickery   
96     Heroes don't always come in uniform   

                                               actors        wins_nominations  \
11  Ranveer Singh|Alia Bhatt|Siddhant Chaturvedi|V...  6 wins & 3 nominations   
34   Gavie Chahal|Mohan Agashe|Mohan Joshi|Lom Harsh|   2 wins & 1 nomination   
37  Ayushmann Khurrana|Nassar|Manoj Pahwa|Kumud Mi...                   1 win   
87  Sidharth Malhotra|Manoj Bajpayee|Rakul Preet S...            1 nomination   
96  Ajay Devgn|Saurabh Shukla|Ileana D'Cruz|Amit S...  2 wins & 3 nominations   

              release_date          lead actor  
11  14 February 2019 (USA)       Ranveer Singh  
34     24 May 2019 (India)        Gavie Chahal  
37      28 June 2019 (USA)  Ayushmann Khurrana  
87  16 February 2018 (USA)   Sidharth Malhotra  
96   16 March 2018 (India)          Ajay Devgn

## Important DataFrame Functions

In [122]:
ipl.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 950 entries, 0 to 949
Data columns (total 20 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   ID               950 non-null    int64  
 1   City             899 non-null    object 
 2   Date             950 non-null    object 
 3   Season           950 non-null    object 
 4   MatchNumber      950 non-null    object 
 5   Team1            950 non-null    object 
 6   Team2            950 non-null    object 
 7   Venue            950 non-null    object 
 8   TossWinner       950 non-null    object 
 9   TossDecision     950 non-null    object 
 10  SuperOver        946 non-null    object 
 11  WinningTeam      946 non-null    object 
 12  WonBy            950 non-null    object 
 13  Margin           932 non-null    float64
 14  method           19 non-null     object 
 15  Player_of_Match  946 non-null    object 
 16  Team1Players     950 non-null    object 
 17  Team2Players    

### Astype
`astype()` changes the dtype of a given column. This can be used to reduce the size of a Dataframe, or for similar purposes.

In [ ]:
# astype
ipl['ID'] = ipl['ID'].astype('int32')
ipl.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 950 entries, 0 to 949
Data columns (total 20 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   ID               950 non-null    int32  
 1   City             899 non-null    object 
 2   Date             950 non-null    object 
 3   Season           950 non-null    object 
 4   MatchNumber      950 non-null    object 
 5   Team1            950 non-null    object 
 6   Team2            950 non-null    object 
 7   Venue            950 non-null    object 
 8   TossWinner       950 non-null    object 
 9   TossDecision     950 non-null    object 
 10  SuperOver        946 non-null    object 
 11  WinningTeam      946 non-null    object 
 12  WonBy            950 non-null    object 
 13  Margin           932 non-null    float64
 14  method           19 non-null     object 
 15  Player_of_Match  946 non-null    object 
 16  Team1Players     950 non-null    object 
 17  Team2Players    

Reduced the Dataframe's size by approximately 3.7 KB. Furthermore, we can reduce more size by changing the dtype of the columns where the same values are repeating or has a very limited number of values to type **category**.

In [126]:
ipl['ID'] = ipl['ID'].astype('int32')
ipl['Season'] = ipl['Season'].astype('category')
ipl['Team1'] = ipl['Team1'].astype('category')
ipl['Team2'] = ipl['Team2'].astype('category')
ipl.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 950 entries, 0 to 949
Data columns (total 20 columns):
 #   Column           Non-Null Count  Dtype   
---  ------           --------------  -----   
 0   ID               950 non-null    int32   
 1   City             899 non-null    object  
 2   Date             950 non-null    object  
 3   Season           950 non-null    category
 4   MatchNumber      950 non-null    object  
 5   Team1            950 non-null    category
 6   Team2            950 non-null    category
 7   Venue            950 non-null    object  
 8   TossWinner       950 non-null    object  
 9   TossDecision     950 non-null    object  
 10  SuperOver        946 non-null    object  
 11  WinningTeam      946 non-null    object  
 12  WonBy            950 non-null    object  
 13  Margin           932 non-null    float64 
 14  method           19 non-null     object  
 15  Player_of_Match  946 non-null    object  
 16  Team1Players     950 non-null    object  
 1